File: python/recursion/binarySearch.py
from typing import List
def binarySearch(arr:List[int], start:int, end:int, element:int)->int or None:
    '''
    returns the position of the search element if found recursively else return None
    input: an integer list: List[int], starting index: int, ending index: int, element to be searched:int
    output: return position of index if found: int or return None
    '''
    if end >= start:
        mid = (start + end) // 2
        guess = arr[mid]
        if guess == element:
            return mid          
        elif guess > element:
            return binarySearch(arr, start, mid -1 , element)
        else:
            return binarySearch(arr, mid+1, end, element)
    else:
        return None    
arr = [1, 2, 3, 4, 5, 6]
print(binarySearch(arr, 0, len(arr) - 1, 1))

File: mycodes/binary_search/binarySearch.cpp
#include<iostream>
#include<vector>
using namespace std;

void input(vector<int>&arr){
    for(auto &i: arr){
        cin >> i;
    }
}

void display(const vector<int>arr){
    for(auto const i: arr){
        cout << i << " ";
    }
    cout << endl;
}

int binarySearch(const vector<int>&arr, int const key){
    //Iterative:-
    int start = 0, end = arr.size() - 1; 
    while (start <= end){
        
        int mid = (start + end )/ 2;
        int guess = arr[mid];
        if (guess == key){
            return mid;
        } 
        if(guess < key){
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }
    return -1;

}
int main(){
    int n;
    cout << "Enter the size of the array: ";
    cin >> n;
    cout << "Enter the elements: ";
    vector<int>arr(n);
    input(arr);
    cout << binarySearch(arr, 2) << endl;
}